src/handlers/applicationFormsHandlers.ts

```diff
@@ -22,27 +22,103 @@ import type {
+    || row.applicationFormFieldPosition === undefined
+    || row.applicationFormFieldPosition === null
+    || row.applicationFormFieldLabel === undefined
+    || row.applicationFormFieldLabel === null
+    || row.applicationFormFieldCreatedAt === undefined
+    || row.applicationFormFieldCreatedAt === null) {
+    throw new Error('Expected id and formField values to be present.');
+  }
+  return {
+    id: row.applicationFormFieldId,
+    applicationFormId: row.id,
+    canonicalFieldId: row.applicationFormFieldCanonicalFieldId,
+    position: row.applicationFormFieldPosition,
+    label: row.applicationFormFieldLabel,
+    createdAt: row.applicationFormFieldCreatedAt,
+  };
+};
+
 const getApplicationForms = (
   req: Request,
   res: Response,
   next: NextFunction,
 ): void => {
   db.sql('applicationForms.selectAll')
-    .then((applicationFormsQueryResult: Result<ApplicationForm>) => {
-      logger.debug(applicationFormsQueryResult);
-      const { rows: applicationForms } = applicationFormsQueryResult;
-      if (isApplicationFormArray(applicationForms)) {
+    .then((applicationFormsQueryResult: Result<ApplicationFormWithFields>) => {
+      const { rows: applicationFormsWithFields } = applicationFormsQueryResult;
+      if (isApplicationFormWithFieldsArray(applicationFormsWithFields)) {
+        let currentApplicationForm: ApplicationForm = {
+          id: 0,
+          opportunityId: 0,
+          version: -1,
+          createdAt: new Date(),
+          fields: [],
+        };
+        const applicationForms: ApplicationForm[] = [];
+        // First, read the ApplicationForm part from this row and compare to
+        // previous row (simply by the id). If it's the same, add only a field.
+        // If it's new, create a new ApplicationForm and add the first field if present.
+        applicationFormsWithFields.forEach((formWithFieldRow) => {
+          if (currentApplicationForm.id === formWithFieldRow.id) {
+            if (formWithFieldRow.applicationFormFieldId !== undefined
+              && formWithFieldRow.applicationFormFieldId !== null) {
+              const formField = getFieldFromRow(formWithFieldRow);
+              // add field to previousRow
+              if (currentApplicationForm.fields === undefined) {
+                currentApplicationForm.fields = [formField];
+              } else {
+                currentApplicationForm.fields.push(formField);
+              }
+            }
+          } else {
+            // Add the previousRow to the response object, then create a new
+            // application form.
+            if (currentApplicationForm.id > 0) {
+              applicationForms.push(currentApplicationForm);
+            }
+            currentApplicationForm = {
+              id: formWithFieldRow.id,
+              opportunityId: formWithFieldRow.opportunityId,
+              version: formWithFieldRow.version,
+              createdAt: formWithFieldRow.createdAt,
+              fields: [],
```

Comment:
Suppose I remove this line 96 and 67 to force that possibility of undefined `fields`, well then a bigger issue arises: the response body might not have the `fields` key at all, which is going to be a problem if you want to consistently be able to parse the response using a schema. It makes the most sense to always have the empty list than to sometimes have it and other times not have it.

```
 FAIL  src/__tests__/applicationForm.int.test.ts (9.288 s)
  ● /applicationForms › GET / › returns all application forms present in the database

    expected [
      {
        createdAt: '2022-07-20T12:00:00.000Z',
        id: 1,
        opportunityId: 1,
        version: 1,
        fields: []
      },
      {
        createdAt: '2022-08-20T12:00:00.000Z',
        id: 2,
        opportunityId: 1,
        version: 2,
        fields: []
      },
      {
        createdAt: '2022-09-20T12:00:00.000Z',
        id: 3,
        opportunityId: 2,
        version: 1,
        fields: []
      }
    ] response body, got [
      {
        id: 1,
        opportunityId: 1,
        version: 1,
        createdAt: '2022-07-20T12:00:00.000Z'
      },
      {
        id: 2,
        opportunityId: 1,
        version: 2,
        createdAt: '2022-08-20T12:00:00.000Z'
      },
      {
        id: 3,
        opportunityId: 2,
        version: 1,
        createdAt: '2022-09-20T12:00:00.000Z'
      }
    ]
```